swarm/storage/pushsync/pusher_test.go

```diff
@@ -0,0 +1,414 @@
+	tagCnt := 4
+
+	errc := make(chan error)
+	sent := &sync.Map{}
+	sendTimes := make(map[int]time.Time)
+	synced := make(map[int]int)
+	quit := make(chan struct{})
+	defer close(quit)
+
+	errf := func(s string, vals ...interface{}) {
+		select {
+		case errc <- fmt.Errorf(s, vals...):
+		case <-quit:
+		}
+	}
+
+	ps := newLoopBack(false)
+
+	max := 0 // the highest index sent so far
+	respond := func(msg []byte, _ *p2p.Peer) error {
+		chmsg, err := decodeChunkMsg(msg)
+		if err != nil {
+			errf("error decoding chunk message: %v", err)
+			return nil
+		}
+		// check outgoing chunk messages
+		cur := int(binary.BigEndian.Uint64(chmsg.Addr[:8]))
+		if cur > max {
+			errf("incorrect order of chunks from db chunk #%d before #%d", cur, max)
+			return nil
+		}
+		v, found := sent.Load(cur)
+		previouslySentAt, repeated := sendTimes[cur]
+		if !found {
+			if !repeated {
+				errf("chunk #%d not sent but received", cur)
+			}
+			return nil
+		}
+		sentAt := v.(time.Time)
+		if repeated {
+			// expect at least retryInterval since previous push
+			if expectedAt := previouslySentAt.Add(retryInterval); expectedAt.After(sentAt) {
+				errf("resync chunk #%d too early. previously sent at %v, next at %v < expected at %v", cur, previouslySentAt, sentAt, expectedAt)
+				return nil
+			}
+		}
+		// remember the latest time sent
+		sendTimes[cur] = sentAt
+		max++
+		// respond ~ mock storer protocol
+		go func() {
+			receipt := &receiptMsg{Addr: chmsg.Addr}
+			rmsg, err := rlp.EncodeToBytes(receipt)
+			if err != nil {
+				errf("error encoding receipt message: %v", err)
+			}
+			log.Debug("chunk sent", "addr", hex.EncodeToString(receipt.Addr))
+			// random delay to allow retries
+			if !delayResponse() {
+				log.Debug("chunk/receipt lost", "addr", hex.EncodeToString(receipt.Addr))
+				return
+			}
+			log.Debug("store chunk,  send receipt", "addr", hex.EncodeToString(receipt.Addr))
+			err = ps.Send(chmsg.Origin, pssReceiptTopic, rmsg)
+			if err != nil {
+				errf("error sending receipt message: %v", err)
+			}
+		}()
+		return nil
+	}
+	// register the respond function
+	ps.Register(pssChunkTopic, false, respond)
+	tags, tagIDs := setupTags(chunkCnt, tagCnt)
+	// construct the mock push sync index iterator
+	tp := newTestPushSyncIndex(chunkCnt, tagIDs, tags, sent)
+	// start push syncing in a go routine
+	p := New(tp, ps, tags)
+	defer p.Close()
+	// collect synced chunks until all chunks synced
+	// wait on errc for errors on any thread
+	// otherwise time out
+	for {
+		select {
+		case i := <-tp.synced:
+			sent.Delete(i)
+			n := synced[i]
+			synced[i] = n + 1
+			if len(synced) == chunkCnt {
+				expTotal := chunkCnt / tagCnt
+				checkTags(t, expTotal, tagIDs[:tagCnt-1], tags)
+				return
+			}
+		case err := <-errc:
+			if err != nil {
+				t.Fatal(err)
+			}
+		case <-time.After(timeout):
+			t.Fatalf("timeout waiting for all chunks to be synced")
+		}
+	}
+
+}
+
+// setupTags constructs tags object create tagCnt - 1 tags
```

Comment:
Why minus one?

Reply:
cos we deliberately keep a tag for which no tag exists